docs/migration.md

```diff
@@ -180,6 +180,58 @@ Model(x=1)
   To work around this, use an `IsInstance` validator (more details to come).
 * Subclasses of built-ins won't validate into their subclass types; you'll need to use an `IsInstance` validator to validate these types.
 
+### Input types are not preserved
+
+In Pydantic V1 we made great efforts to preserve the types of all field inputs for generic collections when they were proper subtypes of the field annotations.
+For example, given the annotation `Mapping[str, int]` if you passed in a `collection.Counter()` you'd get a `collection.Counter()` as the value.
+
+Supporting this behavior in V2 would have negative performance implications for the general case (we'd have to check types every time) and would add a lot of complexity to validation. Further, even in V1 this behavior was inconsistent and partially broken: it did not work for most types (`str`, `UUID`, etc.) and even for generic collections it's impossible to re-build the original input correctly without a lot of special casing (consider `ChainMap`; rebuilding the input is necessary because we need to replace values after validation, e.g. if coercing strings to ints).
+
+In Pydantic V2 we no longer attempt to preserve the input type.
+Instead, we only promise that the output type will match the type annotations.
+Going back to the `Mapping` example, we promise the output will be a valid `Mapping`, in practice it will be a plain `dict`.
+
+```python
+from collections import Counter
+from typing import Mapping
+
+from pydantic import TypeAdapter
+
+print(type(TypeAdapter(Mapping[str, int]).validate_python(Counter())))
+#> <class 'dict'>
+```
+
+If you want the output type to be a specific type, consider annotating it as such or implementing a custom validator:
+
+```python
+from collections import Counter
+from typing import Any, Mapping, TypeVar
+
+from typing_extensions import Annotated
+
+from pydantic import (
+    TypeAdapter,
+    ValidationInfo,
+    ValidatorFunctionWrapHandler,
+    WrapValidator,
+)
+
+
+def restore_input_type(
```

Comment:
better to simply use

```py
ta = TypeAdapter(Counter)
v = ta.validate_python(Counter())
print(type(v))
```

I believe we have specific support for `Counter` - if not I think we should as I believe it was supported in V1.

Reply:
I updated the example to use a custom dict subclass